docs: OpenEBS upgrade guide. (#1163)
kinvolk/lokomotive: docs/how-to-guides/upgrade-openebs.md

# Upgrading and migrating OpenEBS data plane components

## Contents

* [Introduction](#introduction)
* [Prerequisites](#prerequisites)
* [Step 1: Upgrade OpenEBS control plane components](#step-1-upgrade-openebs-control-plane-components)
* [Step 2: Upgrade OpenEBS data plane components](#step-2-upgrade-openebs-data-plane-components)
	* [Upgrade cStor pools](#upgrade-cstor-pools)
	* [Upgrade cStor volumes](#upgrade-cstor-volumes)
* [Step 3: Verify](#step-3-verify)
* [Summary](#summary)
* [Additional resources](#additional-resources)

## Introduction

> This guide provides steps to upgrade OpenEBS control plane and data plane
> components from one version to another.

Reviewer: ```suggestion
This guide provides step-by-step instruction how to upgrade [OpenEBS](https://openebs.io/) control plane and data plane
components from one version to another.
```

> In this document we are going to update from `v1.12.0` to `v2.2.0`.

Reviewer: So we consistently _upgrade_, not _update_.
```suggestion
In this document we are going to upgrade from `v1.12.0` to `v2.2.0`.
```

## Prerequisites

To upgrade the OpenEBS data plane components, we need the following:

* A Kubernetes cluster accessible via `kubectl`.

> * [OpenEBS](../configuration-reference/components/openebs.md) installed. You can

Reviewer: There is no component called `openebs`.
```suggestion
* [OpenEBS component](../configuration-reference/components/openebs-operator.md) installed. You can
```

  check if OpenEBS is indeed in expected version:
    ```bash
    $ kubectl get pods -n openebs -l openebs.io/version=1.12.0
    ```

* Highly recommended to schedule a downtime for the applications consuming the
  OpenEBS PV and make sure to take a backup of the data before starting the

>   below upgrade procedure. Lokomotive provides

Reviewer: Is the disruption expected from the upgrade process? Or may only occur if some problems arise?

Author: Disruption is not expected, only if there is a problem. A precautionary measure.

Reviewer: I'd mention it somehow then. Maybe:
```suggestion
* Upgrade process should not be disruptive, but it is recommended to schedule a downtime for the applications 
  consuming the OpenEBS PV and take a backup of the data before starting the
  below upgrade procedure, in case some problem arises.
  
  Lokomotive provides
```

  [Velero](../configuration-reference/components/velero.md) component for backup
  and restore.

* Ensure the cluster and OpenEBS volumes are is in healthy state before proceeding.
    ```bash
    $ lokoctl health
    Node                    Ready    Reason          Message

    alpha-controller-0      True     KubeletReady    kubelet is posting ready status
    alpha-large-worker-0    True     KubeletReady    kubelet is posting ready status
    Name      Status    Message              Error

    etcd-0    True      {"health":"true"}

   $ kubectl get cstorpools -n openebs # Status should be Healthy.

   NAME                                ALLOCATED   FREE   CAPACITY   STATUS    READONLY   TYPE      AGE
   cstor-pool-openebs-replica-1-w3r7   6.98G       437G   444G       Healthy   false      striped   4d22h

   $ kubectl get cstorvolume -n openebs # Status should be Healthy.

   NAME                                       STATUS    AGE     CAPACITY
   pvc-183dc3df-a7d9-4273-a8e2-7f66d2e19f4e   Healthy   3d20h   50Gi
   pvc-5d4b4c2b-2ed5-4e75-aeb9-fbd2918ebb78   Healthy   3d20h   50Gi
   ```
## Steps

> Set the following environment variables in the terminal to assist in our upgrade:

Reviewer: ```suggestion
Set the following environment variables in the terminal to assist in the upgrade process:
```

```bash

> export OPENEBS_NEW_VERSION=2.2.0
> export OPENEBS_OLD_VERSION=1.12.0

Reviewer: First old, then new. Otherwise it's counter intuitive and may indicate _downgrade_ process.
```suggestion
export OPENEBS_OLD_VERSION=1.12.0
export OPENEBS_NEW_VERSION=2.2.0
```

```
### Step 1: Upgrade OpenEBS control plane components

Lokomotive provides an easy way of upgrading the OpenEBS control plane.

Execute the following command to upgrade OpenEBS control plane:

```bash
lokoctl component apply openebs-operator
```

> Doing so, terminates the OpenEBS resources associated with `v1.12.0` and new
> resources are created with `v2.2.0`.

Reviewer: ```suggestion
Doing so, terminates the OpenEBS resources associated with old version and new
resources are created.
```

Verify all the pods are in `Running` state before proceeding:

```bash
$ kubectl get pods -n openebs -l openebs.io/version=${OPENEBS_NEW_VERSION}
```

### Step 2: Upgrade OpenEBS data plane components

OpenEBS control plane and data plane components work independently. Even if the

> control plane components are upgraded to `v2.2.0`, the data plane components
> continue to work with the older version `v1.12.0`.

Reviewer: ```suggestion
control plane components are upgraded to new version, the data plane components
continue to work with the older version.
```

#### Upgrade cStor pools

Get existing `StoragePoolClaims` by executing:

```bash
kubectl get spc
NAME                           AGE
cstor-pool-openebs-replica-3   2d21h26m
```

Create a Job to upgrade the existing cStor pools:

```bash

> cat > upgrade-cstor-pools-1120-220.yaml <<EOF

Reviewer: BTW, do we need intermediate file here? Maybe we can pipe directly to `kubectl apply` ?

Author: I'd like the user to have a yaml file pre and  post upgrade for reference.

> cat > upgrade-cstor-pools-1120-220.yaml <<EOF
> # upgrade-cstor-pools-1120-220.yaml

Reviewer: ```suggestion
cat > upgrade-cstor-pools-from-${OPENEBS_OLD_VERSION}-to-$(OPENEBS_NEW_VERSION}.yaml <<EOF
```

---
apiVersion: batch/v1
kind: Job
metadata:

>   generateName: cstor-spc-1120220-

Reviewer: This should also use env variables to replace versions.

  namespace: openebs
spec:
  backoffLimit: 4
  template:
    spec:
      serviceAccountName: openebs-operator
      containers:
      - name:  upgrade
        args:

>         - "cstor-spc"
>         - "--from-version=${OPENEBS_OLD_VERSION}"
>         - "--to-version=${OPENEBS_NEW_VERSION}"
> $(kubectl get cstorpools -n openebs --no-headers -o custom-columns=":metadata.name" | sed 's/.*/        - "&"/')
>         - "--v=4"

Reviewer: I'd put verbosity flag at the beginning so literal arguments are at the end.
```suggestion
        - "cstor-spc"
        - "--v=4"
        - "--from-version=${OPENEBS_OLD_VERSION}"
        - "--to-version=${OPENEBS_NEW_VERSION}"
$(kubectl get cstorpools -n openebs --no-headers -o custom-columns=":metadata.name" | sed 's/.*/        - "&"/')
```

        #DO NOT CHANGE BELOW PARAMETERS
        env:
        - name: OPENEBS_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        tty: true
        image: openebs/m-upgrade:${OPENEBS_NEW_VERSION}
        imagePullPolicy: Always
      restartPolicy: OnFailure
EOF
```

Apply the job to start the upgrade for cStor pools:

```bash

> kubectl create -f upgrade-cstor-pools-1120-220.yaml

Reviewer: Ditto hardcoded versions in file names.

```

> Ensure the job runs to completion.

Reviewer: How?

#### Upgrade cStor volumes

Get existing `CStorVolumes` by executing:

```bash
k get cstorvolumes --all-namespaces
NAMESPACE   NAME                                       STATUS    AGE     CAPACITY
openebs     pvc-5b7d8ee3-59b8-4a02-bd6d-a4e66fbecf9f   Healthy   2d21h   50Gi
openebs     pvc-e8f8b268-b81c-4201-841a-283f557c44a7   Healthy   2d21h   50Gi
```

Create a Kubernetes job to upgrade the existing cStor volumes:

```bash

> cat > upgrade-cstor-vols-1120-220.yaml <<EOF

Reviewer: Ditto hardcoded versions and `kubectl apply`.

Author: https://github.com/kinvolk/lokomotive/pull/1163#discussion_r519635065

Reviewer: Hmm. Why use `generateName`? I think this should not be idempotent. What happens if 2 jobs are running simultaneously? I really doubt that would be desired. I think `name` would be safer.

# upgrade-cstor-vols-1120-220.yaml
---
apiVersion: batch/v1
kind: Job
metadata:
  generateName: cstor-vol-1120220-
  namespace: openebs
spec:
  backoffLimit: 4
  template:
    spec:
      serviceAccountName: openebs-operator
      containers:
      - name:  upgrade
        args:
        - "cstor-volume"
        - "--from-version=${OPENEBS_OLD_VERSION}"
        - "--to-version=${OPENEBS_NEW_VERSION}"
$(kubectl get cstorvolumes -n openebs --no-headers -o custom-columns=":metadata.name" | sed 's/.*/        - "&"/')

>         - "--v=4"

Reviewer: Ditto version before arguments.

        #DO NOT CHANGE BELOW PARAMETERS
        env:
        - name: OPENEBS_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        tty: true
        image: openebs/m-upgrade:${OPENEBS_NEW_VERSION}
        imagePullPolicy: Always
      restartPolicy: OnFailure
EOF
```

```bash

> kubectl create -f upgrade-cstor-vols-1120-220.yaml

Reviewer: Ditto versions and `kubectl apply`.

Author: https://github.com/kinvolk/lokomotive/pull/1163#discussion_r519635065

```

> Ensure the job runs to completion.

Reviewer: Ditto how?

# Upgrading and migrating OpenEBS data plane components

## Contents

* [Introduction](#introduction)
* [Prerequisites](#prerequisites)
* [Step 1: Upgrade OpenEBS control plane components](#step-1-upgrade-openebs-control-plane-components)
* [Step 2: Upgrade OpenEBS data plane components](#step-2-upgrade-openebs-data-plane-components)
	* [Upgrade cStor pools](#upgrade-cstor-pools)
	* [Upgrade cStor volumes](#upgrade-cstor-volumes)
* [Step 3: Verify](#step-3-verify)
* [Summary](#summary)
* [Additional resources](#additional-resources)

## Introduction

This guide provides step-by-step instructions on how to upgrade
[OpenEBS](https://openebs.io/) control plane and data plane components from one
version to another.

In this document we are going to upgrade from `v1.12.0` to `v2.2.0`.

## Prerequisites

To upgrade the OpenEBS data plane components, we need the following:

* A Kubernetes cluster accessible via `kubectl`.

* [OpenEBS component](../configuration-reference/components/openebs-operator.md)

>   installed. You can check if OpenEBS is indeed in expected version:

Author: We should keep a document which has the matrix of all the versions installed in the latest version of Lokomotive.

we can then point the user to that matrix to use the correct values ?

>   installed. You can check if OpenEBS is indeed in expected version:

Reviewer: Maybe we can just ask user to check pod labels before/after upgrading the controlplane for now?

>   installed. You can check if OpenEBS is indeed in expected version:

Author: @invidian 
```
## Introduction

This guide provides step-by-step instructions on how to upgrade
[OpenEBS](https://openebs.io/) control plane and data plane components from one
version to another.

To get the current version installed, execute:
```bash
kubectl get pods -n openebs -o jsonpath='{.items[*].metadata.labels.openebs\.io/version}'
``

Refer to the release notes of the Lokomotive release to get the version to
upgrade.

Make note of the current version installed and the version to upgrade and set
the values accordingly in the document.

As an example, in this document we are going to upgrade from `v1.12.0` to
`v2.2.0`.
```

How about this ?

>   installed. You can check if OpenEBS is indeed in expected version:

Reviewer: Sounds good @ipochi !

continue to work with the older version.

#### Upgrade cStor pools

Create a Job to upgrade the existing cStor pools:

```bash
cat > upgrade-cstor-pools-from-${OPENEBS_OLD_VERSION}-to-${OPENEBS_NEW_VERSION}.yaml <<EOF
# upgrade-cstor-pools-from-1.12.0-to-2.2.0.yaml
---
apiVersion: batch/v1
kind: Job
metadata:
  name: cstor-spc-${OPENEBS_OLD_VERSION}-to-${OPENEBS_NEW_VERSION}
  namespace: openebs
spec:
  backoffLimit: 4
  template:
    spec:
      serviceAccountName: openebs-operator
      containers:
      - name:  upgrade
        args:
        - "cstor-spc"
        - "--v=4"
        - "--from-version=${OPENEBS_OLD_VERSION}"
        - "--to-version=${OPENEBS_NEW_VERSION}"
$(kubectl get spc -n openebs --no-headers -o custom-columns=":metadata.name" | sed 's/.*/        - "&"/')
        #DO NOT CHANGE BELOW PARAMETERS
        env:
        - name: OPENEBS_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        tty: true
        image: openebs/m-upgrade:${OPENEBS_NEW_VERSION}
        imagePullPolicy: Always
      restartPolicy: OnFailure
EOF
```

Apply the job to start the upgrade for cStor pools:

```bash

> kubectl apply -f upgrade-cstor-pools-from-${OPENEBS_OLD_VERSION}-to-${OPENEBS_NEW_VERSION}.yaml

Reviewer: If we don't use `generateName`, we can pipe directly to `kubectl apply` right?

Author: https://github.com/kinvolk/lokomotive/pull/1163#discussion_r525841048

openebs-operator-ndm-operator-758fdbc5f4-5qqxx                    1/1     Running   0          34m
openebs-operator-ndm-x44d8                                        1/1     Running   0          34m
openebs-operator-provisioner-59c6dc5dfc-q9k65                     1/1     Running   0          34m
openebs-operator-snapshot-operator-bf49c5dc6-xqqlq                2/2     Running   0          34m
pvc-5b7d8ee3-59b8-4a02-bd6d-a4e66fbecf9f-target-68984f69b7l42sl   3/3     Running   0          22m
pvc-e8f8b268-b81c-4201-841a-283f557c44a7-target-7484b968c4pzld6   3/3     Running   0          20m
```

To check if all the `StoragePoolClaims` and `CStorVolumes` have been upgraded,
execute:

```bash
$ kubectl get pods -n openebs -l openebs.io/version=${OPENEBS_OLD_VERSION}
```
No output should be displayed.

## Summary

This guide helps to upgrade the OpenEBS control plane and data plane components.

## Additional resources

[OpenEBS troubleshooting guide](https://docs.openebs.io/docs/next/troubleshooting.html).

For additional information regarding the upgrade steps, see [OpenEBS upgrade
documentation](https://github.com/openebs/openebs/blob/master/k8s/upgrades/README.md).

> 

Reviewer: Trailing newline can be removed.

---
title: Upgrading OpenEBS control plane and data plane components
weight: 10
---

## Introduction

This guide provides step-by-step instructions on how to upgrade
[OpenEBS](https://openebs.io/) control plane and data plane components from one
version to another.

To get the current version installed, execute:
```bash
kubectl get pods -n openebs -o jsonpath='{.items[*].metadata.labels.openebs\.io/version}'
```

> Refer to the release notes of the Lokomotive release to get the version to

Reviewer: Maybe we can put a link here to the release notes?

Author: We don't have a release notes document if I understand correctly.

Reviewer: We have a changelog, I think we should point there.

Author: ah yes, of course :facepalm:

upgrade.

Make note of the current version installed and the version to upgrade and set
the values accordingly in the document.

As an example, in this document we are going to upgrade from `v1.12.0` to
`v2.2.0`.

## Prerequisites

To upgrade the OpenEBS data plane components, we need the following:

* A Kubernetes cluster accessible via `kubectl`.

* [OpenEBS component](../configuration-reference/components/openebs-operator.md)
  installed. You can check if OpenEBS is indeed in expected version:
    ```bash
    $ kubectl get pods -n openebs -l openebs.io/version=1.12.0
    ```

* Upgrade process should not be disruptive, but it is recommended to schedule a
  downtime for the applications consuming the OpenEBS PV and make sure to take a
  backup of the data before starting the below upgrade procedure in case some
  problem arises. Lokomotive provides

>   [Velero](../configuration-reference/components/velero.md) component for backup

Reviewer: ```suggestion
  [Velero](../../configuration-reference/components/velero) component for backup
```

Links format is now different.

   ```
## Steps

Set the following environment variables in the terminal to assist in the upgrade process:
```bash
export OPENEBS_OLD_VERSION=1.12.0
export OPENEBS_NEW_VERSION=2.2.0
```
### Step 1: Upgrade OpenEBS control plane components

Lokomotive provides an easy way of upgrading the OpenEBS control plane.

Execute the following command to upgrade OpenEBS control plane:

```bash
lokoctl component apply openebs-operator
```

Doing so, terminates the OpenEBS resources associated with the old version and new
resources are created for the new version.

Verify all the pods are in `Running` state before proceeding:

```bash
$ kubectl get pods -n openebs -l openebs.io/version=${OPENEBS_NEW_VERSION}
```

### Step 2: Upgrade OpenEBS data plane components

OpenEBS control plane and data plane components work independently. Even if the
control plane components are upgraded to the new version, the data plane components
continue to work with the older version.

#### Upgrade cStor pools

Create a Job to upgrade the existing cStor pools:

```bash
cat > upgrade-cstor-pools-from-${OPENEBS_OLD_VERSION}-to-${OPENEBS_NEW_VERSION}.yaml <<EOF

> # upgrade-cstor-pools-from-1.12.0-to-2.2.0.yaml

Reviewer 2: Why do we need this line?

Author: No specific need. This line is created when `OPENEBS_OLD_VERSION` and `OPENEBS_NEW_VERSION` is rendered, hence when updating this document this line is not something that would need to be updated manually.

        #DO NOT CHANGE BELOW PARAMETERS
        env:
        - name: OPENEBS_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        tty: true
        image: openebs/m-upgrade:${OPENEBS_NEW_VERSION}
        imagePullPolicy: Always
      restartPolicy: OnFailure
EOF
```

Apply the job to start the upgrade for cStor pools:

```bash
kubectl apply -f upgrade-cstor-pools-from-${OPENEBS_OLD_VERSION}-to-${OPENEBS_NEW_VERSION}.yaml
```

Ensure the job runs to completion:
```bash
kubectl describe job -n openebs cstor-spc-${OPENEBS_OLD_VERSION}-to-${OPENEBS_NEW_VERSION}

# Should see a similar outpu of 'Completed'
  Type    Reason            Age    From            Message
  ----    ------            ----   ----            -------
  Normal  SuccessfulCreate  2m35s  job-controller  Created pod: cstor-spc-1.12.0-to-2.2.0-hpcxl
  Normal  Completed         52s    job-controller  Job completed
```

#### Upgrade cStor volumes

Create a Kubernetes job to upgrade the existing cStor volumes:

```bash
cat > upgrade-cstor-vols-from-${OPENEBS_OLD_VERSION}-to-${OPENEBS_NEW_VERSION}.yaml <<EOF

> # upgrade-cstor-vols-from-1.12.0-to-2.2.0.yaml

Reviewer 2: Why do we need this line?

---
title: Upgrading OpenEBS control plane and data plane components
weight: 10
---

# Upgrading OpenEBS control plane and data plane components

## Contents

* [Introduction](#introduction)
* [Prerequisites](#prerequisites)
* [Step 1: Upgrade OpenEBS control plane components](#step-1-upgrade-openebs-control-plane-components)
* [Step 2: Upgrade OpenEBS data plane components](#step-2-upgrade-openebs-data-plane-components)
	* [Upgrade cStor pools](#upgrade-cstor-pools)
	* [Upgrade cStor volumes](#upgrade-cstor-volumes)
* [Step 3: Verify](#step-3-verify)
* [Summary](#summary)

> * [Additional resources](#additional-resources)

Reviewer: This should be removed in new docs format.

* [Step 2: Upgrade OpenEBS data plane components](#step-2-upgrade-openebs-data-plane-components)
	* [Upgrade cStor pools](#upgrade-cstor-pools)
	* [Upgrade cStor volumes](#upgrade-cstor-volumes)
* [Step 3: Verify](#step-3-verify)
* [Summary](#summary)
* [Additional resources](#additional-resources)

## Introduction

This guide provides installation steps to upgrade OpenEBS dataplane components
from `v1.12.0` to `v2.2.0` and migrate from cStor Pools/Volumes to latest CSI based provisioning
(CSPC Pools/Volumes).

## Prerequisites

To upgrade the OpenEBS data plane components, we need the following:

* Lokomotive `v0.6.0`
    ```bash
    lokoctl version
    v0.6.0
    ```

* A Lokomotive cluster accessible via `kubectl` deployed on a [Equinix]
    metal(https://metal.equinix.com/).

* OpenEBS `v1.12.0` installed. You can check if OpenEBS is indeed in expected
    version:
    ```bash
    $ kubectl get pods -n openebs -n openebs.io/version=1.12.0
    ```

* Highly recommended to schedule a downtime for the applications consuming the
    OpenEBS PV and make sure to take a backup of the data before starting the
    below upgrade procedure.

> 

Reviewer: This has not been addressed.

> 

Author: I added a line 
```
* Ensure the cluster and OpenEBS volumes are is in healthy state before proceeding.
```

I will also add some execution steps as well such as `lokoctl health` and checking the status of OpenEBS volumes.

> 

Reviewer: Cool, LGTM :+1:

#Some of the values below needs to be changed to
#match your openebs installation. The fields are
#indicated with VERIFY
---
apiVersion: batch/v1
kind: Job
metadata:
  #VERIFY that you have provided a unique name for this upgrade job.
  #The name can be any valid K8s string for name. This example uses
  #the following convention: cstor-spc-<flattened-from-to-versions>
  name: cstor-spc-1120220

  #VERIFY the value of namespace is same as the namespace where openebs components
  # are installed. You can verify using the command:
  # `kubectl get pods -n <openebs-namespace> -l openebs.io/component-name=maya-apiserver`
  # The above command should return status of the openebs-apiserver.
  namespace: openebs
spec:
  backoffLimit: 4
  template:
    spec:
      serviceAccountName: openebs-operator
      containers:
      - name:  upgrade
        args:
        - "cstor-spc"

        # --from-version is the current version of the pool
        - "--from-version=1.12.0"

        # --to-version is the version desired upgrade version
        - "--to-version=2.2.0"

        # Bulk upgrade is supported
        # To make use of it, please provide the list of SPCs
        # as mentioned below

>         - "cstor-pool-openebs-replica-3"

Reviewer: Same for storage pools, we should also be able to automatically obtain this list I think.

Author: I don't want to complicate the upgrade steps.

Reviewer: IMO it would simplify it.

Author: @invidian automatically obtain this list and then how to render ?

Reviewer: we already use `cat << EOF` here. we can just replace this with `$()`.

Author: ok but then how to get it listed in the below format without reducing readibility.
```
- "cstor-pool-openebs-replica-3"
- "xyz"
- "abc"
```

Author: @invidian this works :see_no_evil: but I am not sure if I want that.
```
OPENEBS_OLD_VERSION=1.12.0
OPENEBS_NEW_VERSION=2.2.0
cat > upgrade-cstor-pools-1120-220.yaml <<EOF
# upgrade-cstor-pools-1120-220.yaml

#This is an example YAML for upgrading cstor SPC.
#Some of the values below needs to be changed to
#match your openebs installation. The fields are
#indicated with VERIFY
---
apiVersion: batch/v1
kind: Job
metadata:
  #VERIFY that you have provided a unique name for this upgrade job.
  #The name can be any valid K8s string for name. This example uses
  #the following convention: cstor-spc-<flattened-from-to-versions>
  generateName: cstor-spc-1120220-
  namespace: openebs

spec:
  backoffLimit: 4
  template:
    spec:
      serviceAccountName: openebs-operator
      containers:
      - name:  upgrade
        args:
        - "cstor-spc"

        # --from-version is the current version of the pool
        - "--from-version=${OPENEBS_OLD_VERSION}"

        # --to-version is the version desired upgrade version
        - "--to-version=${OPENEBS_NEW_VERSION}"

        # Bulk upgrade is supported
        # To make use of it, please provide the list of SPCs
        # as mentioned below
        - "cstor-pool-openebs-replica-3"
$(kubectl get cstorvolumes -n openebs | awk '{if(NR>1) printf("\"%s\"\n", $1);}'| sed 's/./        - &/')

        #Following are optional parameters
        #Log Level
        - "--v=4"
        #DO NOT CHANGE BELOW PARAMETERS
        env:
        - name: OPENEBS_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        tty: true

        # the image version should be same as the --to-version mentioned above
        # in the args of the job
        image: openebs/m-upgrade:${OPENEBS_NEW_VERSION}
        imagePullPolicy: Always
      restartPolicy: OnFailure
EOF
```

Reviewer: Looks okay, but maybe use `cut` or `awk '{print $1}'` with `tr \\n ' '` or something to make it easier to read? I can help if you give me raw output from `kubectl get cstorvolumes -n openebs` :)

Or even better, use `kubectl get cstorvolumes -n openebs --no-headers -o custom-columns=":metadata.name" | sed 's/^/        - /'`

        #Following are optional parameters
        #Log Level
        - "--v=4"
        #DO NOT CHANGE BELOW PARAMETERS
        env:
        - name: OPENEBS_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        tty: true

        # the image version should be same as the --to-version mentioned above
        # in the args of the job
        image: openebs/m-upgrade:2.2.0
        imagePullPolicy: Always
      restartPolicy: OnFailure

> ---

Reviewer: Not needed line.

# Upgrading and migrating OpenEBS data plane components

## Contents

* [Introduction](#introduction)
* [Prerequisites](#prerequisites)
* [Step 1: Upgrade OpenEBS control plane components](#step-1-upgrade-openebs-control-plane-components)
* [Step 2: Upgrade OpenEBS data plane components](#step-2-upgrade-openebs-data-plane-components)
	* [Upgrade cStor pools](#upgrade-cstor-pools)
	* [Upgrade cStor volumes](#upgrade-cstor-volumes)
* [Step 3: Verify](#step-3-verify)
* [Summary](#summary)
* [Additional resources](#additional-resources)

## Introduction

This guide provides steps to upgrade OpenEBS dataplane components from `v1.12.0`
to `v2.2.0` and migrate from cStor Pools/Volumes to latest CSI based
provisioning (CSPC Pools/Volumes).

## Prerequisites

To upgrade the OpenEBS data plane components, we need the following:

* Lokomotive `v0.6.0`
    ```bash
    lokoctl version
    v0.6.0
    ```

* A Lokomotive cluster accessible via `kubectl` deployed on a
  [Equinix metal](../configuration-reference/platforms/packet.md) or

>   [Bare metal](../configuration-reference/platforms/baremetal.md).

Reviewer: What if I deploy `OpenEBS` component to non-Lokomotive cluster? :smile:

Author: This guide is not for non Lokomotive clusters.

Reviewer: Hmm, why? I think we can just say `openebs` component installed. As far as I know, it's completely fine to use components on non-Lokomotive clusters.

  [Velero](../configuration-reference/components/velero.md) component for backup
  and restore.

* Ensure the cluster and OpenEBS volumes are is in healthy state before proceeding.

## Steps

### Step 1: Upgrade OpenEBS control plane components

Lokomotive provides an easy way of upgrading the OpenEBS control plane.

Execute the following command to upgrade OpenEBS control plane:

```bash
lokoctl component apply openebs-operator
```

Doing so, terminates the OpenEBS resources associated with `v1.12.0` and new
resources are created with `v2.2.0`.

Verify all the pods are in `Running` state before proceeding:

```bash
$ kubectl get pods -n openebs -n openebs.io/version=2.2.0
```

### Step 2: Upgrade OpenEBS data plane components

OpenEBS control plane and data plane components work independently. Even if the
control plane components are upgraded to `v2.2.0`, the data plane components
continue to work with the older version `v1.12.0`.

#### Upgrade cStor pools

Get existing `StoragePoolClaims` by executing:

```bash
kubectl get spc
NAME                           AGE
cstor-pool-openebs-replica-3   2d21h26m
```

Create a Job to upgrade the existing cStor pools:

```bash

> OPENEBS220=2.2.0

Reviewer: I doubt @surajssd will like `OPENEBS220` variable name, I don't think this is what he meant :smile: I think we should have something like `OLD_VERSION` and `NEW_VERSION` or something.

Author: `OPENEBS_NEW_VERSION` and `OPENEBS_OLD_VERSION`

#match your openebs installation. The fields are
#indicated with VERIFY
---
apiVersion: batch/v1
kind: Job
metadata:
  #VERIFY that you have provided a unique name for this upgrade job.
  #The name can be any valid K8s string for name. This example uses
  #the following convention: cstor-spc-<flattened-from-to-versions>
  name: cstor-spc-1120220

  #VERIFY the value of namespace is same as the namespace where openebs components
  # are installed. You can verify using the command:
  # `kubectl get pods -n <openebs-namespace> -l openebs.io/component-name=maya-apiserver`
  # The above command should return status of the openebs-apiserver.
  namespace: openebs
spec:
  backoffLimit: 4
  template:
    spec:
      #VERIFY the value of serviceAccountName is pointing to service account
      # created within openebs namespace. Use the non-default account.
      # by running `kubectl get sa -n <openebs-namespace>`
      serviceAccountName: openebs-operator
      containers:
      - name:  upgrade
        args:
        - "cstor-spc"

        # --from-version is the current version of the pool
        - "--from-version=1.12.0"

        # --to-version is the version desired upgrade version

>         - "--to-version=2.2.0"

Reviewer 2: I would like the user to do less manual steps of editing versions. I think the best way to do is using env vars and tools like `envsubst` or simple bash foo with `echo` also works fine.

        # in the args of the job
        image: openebs/m-upgrade:2.2.0
        imagePullPolicy: Always
      restartPolicy: OnFailure
---
```

Apply the job to start the upgrade for cStor pools:

```bash
kubectl apply upgrade-cstor-pools-1120-220.yaml
```

Ensure the job runs to completion.

#### Upgrade cStor volumes

Get existing `CStorVolumes` by executing:

```bash
k get cstorvolumes --all-namespaces
NAMESPACE   NAME                                       STATUS    AGE     CAPACITY
openebs     pvc-5b7d8ee3-59b8-4a02-bd6d-a4e66fbecf9f   Healthy   2d21h   50Gi
openebs     pvc-e8f8b268-b81c-4201-841a-283f557c44a7   Healthy   2d21h   50Gi
```

Create a Kubernetes job to upgrade the existing cStor volumes:

```bash
# upgrade-cstor-vols-1120-220.yaml

#This is an example YAML for upgrading cstor volume.
#Some of the values below needs to be changed to
#match your openebs installation. The fields are
#indicated with VERIFY
---
apiVersion: batch/v1
kind: Job
metadata:
  #VERIFY that you have provided a unique name for this upgrade job.
  #The name can be any valid K8s string for name. This example uses
  #the following convention: cstor-vol-<flattened-from-to-versions>

>   name: cstor-vol-1120220

Reviewer 2: also here a little kubectl magic with bash should help us reduce manual edits.

Author: I've changed the `name` to `generateName`. Since `generateName` cannot work with `kubectl apply`, the resulting command is also changed to `kubectl create`, which is correct IMO as the intent is to create a job.

# Upgrading and migrating OpenEBS data plane components

## Contents

* [Introduction](#introduction)
* [Prerequisites](#prerequisites)
* [Step 1: Upgrade OpenEBS control plane components](#step-1-upgrade-openebs-control-plane-components)
* [Step 2: Upgrade OpenEBS data plane components](#step-2-upgrade-openebs-data-plane-components)
	* [Upgrade cStor pools](#upgrade-cstor-pools)
	* [Upgrade cStor volumes](#upgrade-cstor-volumes)
* [Step 3: Verify](#step-3-verify)
* [Summary](#summary)
* [Additional resources](#additional-resources)

## Introduction

> This guide provides steps to upgrade OpenEBS dataplane components from `v1.12.0`
> to `v2.2.0` and migrate from cStor Pools/Volumes to latest CSI based
> provisioning (CSPC Pools/Volumes).

Reviewer: ```suggestion
This guide provides steps to upgrade OpenEBS dataplane components from one version to another
and migrate from cStor Pools/Volumes to latest CSI based provisioning (CSPC Pools/Volumes).

For specific example, we will be updating from `v1.12.0` to `v2.2.0`.
```

## Prerequisites

To upgrade the OpenEBS data plane components, we need the following:

* Lokomotive `v0.6.0`
    ```bash
    lokoctl version
    v0.6.0
    ```

* A Lokomotive cluster accessible via `kubectl` deployed on a
  [Equinix metal](../configuration-reference/platforms/packet.md) or
  [Bare metal](../configuration-reference/platforms/baremetal.md).

> * OpenEBS `v1.12.0` installed. You can check if OpenEBS is indeed in expected

Reviewer: ```suggestion
* OpenEBS [component](link-to-component) installed. You can check if OpenEBS is indeed in expected
```

    version:
    ```bash

>     $ kubectl get pods -n openebs -n openebs.io/version=1.12.0

Reviewer 2: `s | -n openebs.io | -l openebs.io | g`

   pvc-183dc3df-a7d9-4273-a8e2-7f66d2e19f4e   Healthy   3d20h   50Gi
   pvc-5d4b4c2b-2ed5-4e75-aeb9-fbd2918ebb78   Healthy   3d20h   50Gi
   ```
## Steps

### Step 1: Upgrade OpenEBS control plane components

Lokomotive provides an easy way of upgrading the OpenEBS control plane.

Execute the following command to upgrade OpenEBS control plane:

```bash
lokoctl component apply openebs-operator
```

Doing so, terminates the OpenEBS resources associated with `v1.12.0` and new
resources are created with `v2.2.0`.

Verify all the pods are in `Running` state before proceeding:

```bash
$ kubectl get pods -n openebs -n openebs.io/version=2.2.0
```

### Step 2: Upgrade OpenEBS data plane components

OpenEBS control plane and data plane components work independently. Even if the
control plane components are upgraded to `v2.2.0`, the data plane components
continue to work with the older version `v1.12.0`.

#### Upgrade cStor pools

Get existing `StoragePoolClaims` by executing:

```bash
kubectl get spc
NAME                           AGE
cstor-pool-openebs-replica-3   2d21h26m
```

Create a Job to upgrade the existing cStor pools:

```bash
OPENEBS_NEW_VERSION=2.2.0

> OPENEBS_OLD_VERSION=1.12.0

Reviewer: I'd set those at the very beginning and re-use through entire guide. Also they should be swapped:
```
OPENEBS_OLD_VERSION=1.12.0
OPENEBS_NEW_VERSION=2.2.0
```

As this order may indicate _downgrading_, not updating.

cat > upgrade-cstor-pools-1120-220.yaml <<EOF
# upgrade-cstor-pools-1120-220.yaml

#This is an example YAML for upgrading cstor SPC.
#Some of the values below needs to be changed to
#match your openebs installation. The fields are
#indicated with VERIFY
---
apiVersion: batch/v1
kind: Job
metadata:
  #VERIFY that you have provided a unique name for this upgrade job.
  #The name can be any valid K8s string for name. This example uses
  #the following convention: cstor-spc-<flattened-from-to-versions>
  generateName: cstor-spc-1120220-
  namespace: openebs

spec:
  backoffLimit: 4
  template:
    spec:
      serviceAccountName: openebs-operator
      containers:
      - name:  upgrade
        args:
        - "cstor-spc"

>         # --from-version is the current version of the pool

Reviewer 2: we can get rid of all the comments in the files to make it concise. Also it is understandable without comments what a particular flag does because we have those env vars which are very descriptive.

        # as mentioned below
        - "cstor-pool-openebs-replica-3"

        #Following are optional parameters
        #Log Level
        - "--v=4"
        #DO NOT CHANGE BELOW PARAMETERS
        env:
        - name: OPENEBS_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        tty: true

        # the image version should be same as the --to-version mentioned above
        # in the args of the job
        image: openebs/m-upgrade:${OPENEBS_NEW_VERSION}
        imagePullPolicy: Always
      restartPolicy: OnFailure
EOF
```

Apply the job to start the upgrade for cStor pools:

```bash
kubectl create -f upgrade-cstor-pools-1120-220.yaml
```

Ensure the job runs to completion.

#### Upgrade cStor volumes

Get existing `CStorVolumes` by executing:

```bash
k get cstorvolumes --all-namespaces
NAMESPACE   NAME                                       STATUS    AGE     CAPACITY
openebs     pvc-5b7d8ee3-59b8-4a02-bd6d-a4e66fbecf9f   Healthy   2d21h   50Gi
openebs     pvc-e8f8b268-b81c-4201-841a-283f557c44a7   Healthy   2d21h   50Gi
```

Create a Kubernetes job to upgrade the existing cStor volumes:

```bash
OPENEBS_NEW_VERSION=2.2.0

> OPENEBS_OLD_VERSION=1.12.0

Reviewer: Ditto set at the beginning.

Author: I want to set it where I am going to use it in the cat EOF script. It wouldn't make sense at the beginning of the document where its not going to be rendered.

Reviewer: I think it's OK to assume they are set and everything happens in the same terminal session. We do it in other tutorials. IMO re-assigning is just confusing.

openebs-operator-admission-server-6b5fb6dff5-6dc8d                1/1     Running   2          34m
openebs-operator-apiserver-5c467bc588-fsl58                       1/1     Running   0          34m
openebs-operator-localpv-provisioner-74d76d55b-s8zwv              1/1     Running   0          33m
openebs-operator-ndm-cfnxp                                        1/1     Running   0          34m
openebs-operator-ndm-nsqkg                                        1/1     Running   0          33m
openebs-operator-ndm-operator-758fdbc5f4-5qqxx                    1/1     Running   0          34m
openebs-operator-ndm-x44d8                                        1/1     Running   0          34m
openebs-operator-provisioner-59c6dc5dfc-q9k65                     1/1     Running   0          34m
openebs-operator-snapshot-operator-bf49c5dc6-xqqlq                2/2     Running   0          34m
pvc-5b7d8ee3-59b8-4a02-bd6d-a4e66fbecf9f-target-68984f69b7l42sl   3/3     Running   0          22m
pvc-e8f8b268-b81c-4201-841a-283f557c44a7-target-7484b968c4pzld6   3/3     Running   0          20m
```

To check if all the `StoragePoolClaims` and `CStorVolumes` have been upgraded,
execute:

```bash

> $ kubectl get pods -n openebs -n openebs.io/version=1.12.0

Reviewer: Use variable instead.

NAME                           AGE
cstor-pool-openebs-replica-3   2d21h26m
```

Create a Job to upgrade the existing cStor pools:

```bash
# upgrade-cstor-pools-1120-220.yaml

#This is an example YAML for upgrading cstor SPC.
#Some of the values below needs to be changed to
#match your openebs installation. The fields are
#indicated with VERIFY
---
apiVersion: batch/v1
kind: Job
metadata:
  #VERIFY that you have provided a unique name for this upgrade job.
  #The name can be any valid K8s string for name. This example uses
  #the following convention: cstor-spc-<flattened-from-to-versions>
  name: cstor-spc-1120220

  #VERIFY the value of namespace is same as the namespace where openebs components
  # are installed. You can verify using the command:
  # `kubectl get pods -n <openebs-namespace> -l openebs.io/component-name=maya-apiserver`
  # The above command should return status of the openebs-apiserver.
  namespace: openebs
spec:
  backoffLimit: 4
  template:
    spec:
      #VERIFY the value of serviceAccountName is pointing to service account
      # created within openebs namespace. Use the non-default account.
      # by running `kubectl get sa -n <openebs-namespace>`

>       serviceAccountName: openebs-operator

Reviewer 2: I think this comment can be removed since we know what the service account is called.

Reviewer 2: Or as lokomotive we control what it is called.
